src/base.jl

```diff
@@ -156,169 +157,105 @@ This is a temporary implementation that will allow consumers of PowerSystems to
 functionality before it is finalized.
 """
 struct ConcreteSystem <: PowerSystemType
-    data::Dict{DataType, Vector{<:Component}}  # Contains arrays of concrete types.
-    components::Dict{DataType, Any}             # Nested dict based on type hierarchy
-                                                # containing references to component arrays.
-    basepower::Float64                          # [MVA]
+    components::Dict{DataType, Vector{<:Component}}    # Contains arrays of concrete types.
+    forecasts::Union{Nothing, SystemForecasts}
+    basepower::Float64                                 # [MVA]
 end
 
 function ConcreteSystem(sys::System)
-    data = Dict{DataType, Vector{<:Component}}()
-    for subtype in get_all_concrete_subtypes(Component)
-        data[subtype] = Vector{subtype}()
-    end
-
-    @debug "Created data keys" keys(data)
+    components = Dict{DataType, Vector{<:Component}}()
+    concrete_sys = ConcreteSystem(components, sys.forecasts, sys.basepower)
 
     for field in (:buses, :loads)
-        objs = getfield(sys, field)
-        for obj in objs
-            push!(data[typeof(obj)], obj)
+        for obj in getfield(sys, field)
+            add_component(concrete_sys, obj)
         end
     end
 
     for field in (:thermal, :renewable, :hydro)
         generators = getfield(sys.generators, field)
         if !isnothing(generators)
             for gen in generators
-                push!(data[typeof(gen)], gen)
+                add_component(concrete_sys, gen)
             end
         end
     end
 
-    for field in (:branches, :storage)
+    for field in (:branches, :storage, :services)
         objs = getfield(sys, field)
         if !isnothing(objs)
             for obj in objs
-                push!(data[typeof(obj)], obj)
+                add_component(concrete_sys, obj)
             end
         end
     end
 
-    for (key, value) in data
-        @debug "data: $(string(key)): count=$(string(length(value)))"
+    for (key, value) in concrete_sys.components
+        @debug "components: $(string(key)): count=$(string(length(value)))"
     end
 
-    components = _get_components_by_type(Component, data)
-    return ConcreteSystem(data, components, sys.basepower)
+    return concrete_sys
 end
 
-"""Returns an array of components from the System. T must be a concrete type.
-
-# Example
-```julia
-devices = PowerSystems.get_components(ThermalDispatch, system)
-```
-"""
-function get_components(::Type{T}, sys::ConcreteSystem)::Vector{T} where {T <: Component}
+"""Adds a component to the system."""
+function add_component(sys::ConcreteSystem, component::T) where T <: Component
```

Comment:
I should have called this add_component!  Do you want to make the change?  Or I can do it in a subsequent commit.

Reply:
Done